### PR TITLE
[feat][development] vernemq certificate acl integration

### PR DESCRIPTION
**Pull request by jonaphael**
* **Please check if the PR fulfills these requirements**
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Is there any issue related to this PR? (Link to an issue, e.g. #99999)** 

* **Other information**:

### CODE REVIEW

File: connector/mqtt/vernemq/broker/vmq_ssl.erl [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+            Subject = TBSCert#'OTPTBSCertificate'.subject,
+            
+            FingerPrint = crypto:hash(sha256, Cert),
+            % the fingerprint at this point is with formatted but with the ":" as the last character
+            FingerPrintFormatted = lists:flatten([io_lib:format("~2.16.0B:",[X]) || <<X:8>> <= FingerPrint ]),
+            ValidFingerPrint = [string:strip(FingerPrintFormatted, right, $:)],
+
+            Username = extract_cn(Subject), 
+            {ok, Client} = eredis:start_link("vernemq-redis", 6379,0),            
+            {ok, ValueReturn} = eredis:q(Client, ["GET", ValidFingerPrint]),
```
Comment (rascaraficci) [inline connector/mqtt/vernemq/broker/vmq_ssl.erl:45]: ValueReturned?

File: connector/mqtt/vernemq/broker/vmq_ssl.erl [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+            eredis:stop(Client),
+           
+            OkReturn = check_fingerprint_username(Username, ValueReturn),
```
Comment (rascaraficci) [inline connector/mqtt/vernemq/broker/vmq_ssl.erl:48]: The Username and ValueReturn are not necessary equals. We removed the rule that the certificate CNAME must meet the pattern `tenant:deviceId`. I think it is only necessary to check if the fingerprint exists in the Redis database.

File: connector/mqtt/vernemq/broker/vmq_ssl.erl [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+socket_to_common_name(Socket) ->
+    case ssl:peercert(Socket) of
+        {error, no_peercert} ->
+            undefined;
+        {ok, Cert} ->
+            OTPCert = public_key:pkix_decode_cert(Cert, otp),
+            TBSCert = OTPCert#'OTPCertificate'.tbsCertificate,
+            Subject = TBSCert#'OTPTBSCertificate'.subject,
+            
+            FingerPrint = crypto:hash(sha256, Cert),
```
Comment (rascaraficci) [inline connector/mqtt/vernemq/broker/vmq_ssl.erl:38]: Is it possible to add a flag to indicate if the fingerprint should be checked or not?

**Review by rascaraficci — COMMENTED** [review]
Good job! I left few comments. I think we are in the right way!

File: connector/mqtt/vernemq/broker/vmq_ssl.erl [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    case Cert of
+        undefined ->
+            undefined;
+        _ ->
+            OTPCert = public_key:pkix_decode_cert(Cert, otp),
+            TBSCert = OTPCert#'OTPCertificate'.tbsCertificate,
+            Subject = TBSCert#'OTPTBSCertificate'.subject,
+            
+            % error_logger:info_msg("Client ID1: ~p ~n", [crypto:hash(sha256, Cert)]),
+            error_logger:info_msg("SHOULD CALL REDIS HERE"),
```
Comment (thiagodpf) [inline connector/mqtt/vernemq/broker/vmq_ssl.erl:63]: >> "SHOULD CALL REDIS HERE"

Is it an implementation for a future feature?

**Review by thiagodpf — APPROVED** [review]

**Review by rascaraficci — APPROVED** [review]
It looks good for a first version.

**Comment by bizetti** [comment]
#1938

**Pull request merged**